source/base/function_spherical.cc

```diff
@@ -0,0 +1,303 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 2016 by the deal.II authors
+//
+// This file is part of the deal.II library.
+//
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE at
+// the top level of the deal.II distribution.
+//
+// ---------------------------------------------------------------------
+
+#include <deal.II/base/point.h>
+#include <deal.II/base/function_spherical.h>
+#include <deal.II/base/geometric_utilities.h>
+
+#include <cmath>
+#include <algorithm>
+
+DEAL_II_NAMESPACE_OPEN
+namespace Functions
+{
+
+  // other implementations/notes:
+  // https://github.com/apache/commons-math/blob/master/src/main/java/org/apache/commons/math4/geometry/euclidean/threed/SphericalCoordinates.java
+  // http://mathworld.wolfram.com/SphericalCoordinates.html
+
+  /*derivation of Hessian in Maxima as function fo tensor products of unit vectors:
+
+  depends(ur,[theta,phi]);
+  depends(utheta,theta);
+  depends(uphi,[theta,phi]);
+  depends(f,[r,theta,phi]);
+  declare([f,r,theta,phi], scalar)$
+  dotscrules: true;
+  grads(a):=ur.diff(a,r)+(1/r)*uphi.diff(a,phi)+(1/(r*sin(phi)))*utheta.diff(a,theta);
+
+
+  H : factor(grads(grads(f)));
+  H2: subst([diff(ur,theta)=sin(phi)*utheta,
+       diff(utheta,theta)=-cos(phi)*uphi-sin(phi)*ur,
+       diff(uphi,theta)=cos(phi)*utheta,
+       diff(ur,phi)=uphi,
+       diff(uphi,phi)=-ur],
+      H);
+  H3: fullratsimp(H2);
+
+
+  H4 : subst([diff(f,r)=sg0,
+            diff(f,theta)=sg1,
+            diff(f,phi)=sg2,
+            diff(f,r,2)=sh00,
+            diff(f,theta,2)=sh11,
+            diff(f,phi,2)=sh22,
+            diff(f,r,1,theta,1)=sh01,
+            diff(f,r,1,phi,1)=sh02,
+            diff(f,theta,1,phi,1)=sh12,
+            cos(phi)=cos_phi,
+            cos(theta)=cos_theta,
+            sin(phi)=sin_phi,
+            sin(theta)=sin_theta
+            ],
+      H3);
+
+  where (used later to do tensor products):
+
+  ur     : [cos(theta)*sin(phi), sin_theta*sin(phi), cos_phi];
+  utheta : [-sin_theta, cos(theta), 0];
+  uphi   : [cos(theta)*cos_phi, sin_theta*cos_phi, -sin(phi)];
+
+   */
+
+  namespace
+  {
+    /**
+     * calculates out[i][j] = in1[i]in2[j]+in1[j]in2[i]
+     */
+    template <int dim>
+    void symm_outer_product(SymmetricTensor<2,dim> &out,
+                            const Tensor<1,dim> &in1,
+                            const Tensor<1,dim> &in2)
```

Comment:
By the way, that's a funny name because the function is not the same as `symmetrize(outer_product(in1, in2))`. The latter will divide the result by two.